### PR TITLE
XCOREVBITMANIP builtins added

### PR DESCRIPTION
**Pull request by MaryBennett**
Files Changed:
gcc/common/config/riscv:
 * riscv-common.cc: Added the XCOREVBITMANIP extension.

gcc/config/riscv:
 * constraints.md: Added builtins for the XCOREVBITMANIP extension.
 * corev.def: Likewise.
 * corev.md: Likewise.
 * predicates.md: Likewise.
 * riscv-builtins.cc: Likewise.
 * riscv-ftypes.def: Likewise.
 * riscv-opts.h: Likewise.
 * riscv.md: Likewise.

gcc/doc:
 * extend.texi: Added documentation for the XCOREVBITMANIP extension builtins.

gcc/testsuite/gcc.target/riscv:
 * cv-march-xcorevbitmanip-compile-bclr.c: Added new tests.
 * cv-march-xcorevbitmanip-compile-bclrr.c: Likewise.
 * cv-march-xcorevbitmanip-compile-bitrev.c: Likewise.
 * cv-march-xcorevbitmanip-compile-bset.c: Likewise.
 * cv-march-xcorevbitmanip-compile-bsetr.c: Likewise.
 * cv-march-xcorevbitmanip-compile-clb.c: Likewise.
 * cv-march-xcorevbitmanip-compile-cnt.c: Likewise.
 * cv-march-xcorevbitmanip-compile-extract.c: Likewise.
 * cv-march-xcorevbitmanip-compile-extractr.c: Likewise.
 * cv-march-xcorevbitmanip-compile-extractu.c: Likewise.
 * cv-march-xcorevbitmanip-compile-extractur.c: Likewise.
 * cv-march-xcorevbitmanip-compile-ff1.c: Likewise.
 * cv-march-xcorevbitmanip-compile-fl1.c: Likewise.
 * cv-march-xcorevbitmanip-compile-insert.c: Likewise.
 * cv-march-xcorevbitmanip-compile-insertr.c: Likewise.
 * cv-march-xcorevbitmanip-compile-ror.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-bclr.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-bitrev.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-bset.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-extract.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-extractu.c: Likewise.
 * cv-march-xcorevbitmanip-fail-compile-insert.c: Likewise.

gcc/testsuite/lib:
 * target-supports.exp: Added proc for the XCOREVBITMANIP extension.

==Results==

| Category             | Previous | With commit | Delta |
| -------------------: | -------: | ----------: | ----: |
| Expected passes      |     3681 |        4299 |     +618 |
| Unexpected failures  |       53 |          53 |     - | 
| Unexpected successes |        - |           - |     - | 
| Expected failures    |        6 |           6 |     - | 
| Unresolved testcases |        6 |           6 |     - | 
| Unsupported tests    |      233 |         245 |    +12 |

Similar to the tests for the hardware loop extension, failure tests using the ```lto``` optimisation are marked as unsupported.

### CODE REVIEW

File: gcc/config/riscv/constraints.md [hunk 1]
```diff
@@ -80,3 +80,14 @@
    A constant @code{move_operand}."
   (and (match_operand 0 "move_operand")
        (match_test "CONSTANT_P (op)")))
+
+;; CORE-V Constraints
+(define_constraint "M"
```
Comment (jeremybennett) [inline gcc/config/riscv/constraints.md:85]: I think these might need a more CORE-V specific name. Possilby "CV10B" for "M" and "CV2B" for N

File: gcc/config/riscv/corev.def [hunk 2]
```diff
@@ -20,3 +20,16 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
 RISCV_BUILTIN (cv_mac_machhuRN,  "cv_mac_machhuRN",   RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI_USI_UQI,  cvmac),
 RISCV_BUILTIN (cv_mac_macsRN,    "cv_mac_macsRN",     RISCV_BUILTIN_DIRECT, RISCV_SI_FTYPE_SI_SI_SI_UQI,      cvmac),
 RISCV_BUILTIN (cv_mac_machhsRN,  "cv_mac_machhsRN",   RISCV_BUILTIN_DIRECT, RISCV_SI_FTYPE_SI_SI_SI_UQI,      cvmac),
+
+// xcorevbitmanip
+RISCV_BUILTIN (cv_bitmanip_extract,     "cv_bitmanip_extract",     RISCV_BUILTIN_DIRECT, RISCV_SI_FTYPE_SI_UHI,          cvbitmanip),
```
Comment (jeremybennett) [inline gcc/config/riscv/corev.def]: I think this is correct, but it highlights a bug in the [spec for cv_bitmanip_extract](https://github.com/openhwgroup/core-v-sw/blob/master/specifications/corev-builtin-spec.md#int32_t-__builtin_riscv_cv_bitmanip_extract-uint32_t--i-uint16_t-range) which specifies the first argument as `uint32_t` (I think `int32_t` makes more sense).

File: gcc/config/riscv/corev.def [hunk 3]
```diff
@@ -20,3 +20,16 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
...
+RISCV_BUILTIN (cv_bitmanip_extractu,    "cv_bitmanip_extractu",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_UHI,        cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_insert,      "cv_bitmanip_insert",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_UHI_USI,    cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_bclr,        "cv_bitmanip_bclr",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_UHI_USI,    cvbitmanip),
```
Comment (jeremybennett) [inline gcc/config/riscv/corev.def]: Surely this should be `RISCV_USI_FTYPE_USI_UHI`

File: gcc/config/riscv/corev.def [hunk 4]
```diff
@@ -20,3 +20,16 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
...
+RISCV_BUILTIN (cv_bitmanip_bset,        "cv_bitmanip_bset",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_UHI_USI,    cvbitmanip),
```
Comment (jeremybennett) [inline gcc/config/riscv/corev.def]: Surely this should be `RISCV_USI_FTYPE_USI_UHI`

File: gcc/config/riscv/corev.def [hunk 5]
```diff
@@ -20,3 +20,16 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
...
+RISCV_BUILTIN (cv_bitmanip_ff1,         "cv_bitmanip_ff1",         RISCV_BUILTIN_DIRECT, RISCV_UQI_FTYPE_USI,            cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_fl1,         "cv_bitmanip_fl1",         RISCV_BUILTIN_DIRECT, RISCV_UQI_FTYPE_USI,            cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_clb,         "cv_bitmanip_clb",         RISCV_BUILTIN_DIRECT, RISCV_UQI_FTYPE_USI,            cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_cnt,         "cv_bitmanip_cnt",         RISCV_BUILTIN_DIRECT, RISCV_UQI_FTYPE_USI,            cvbitmanip),
+RISCV_BUILTIN (cv_bitmanip_ror,         "cv_bitmanip_ror",         RISCV_BUILTIN_DIRECT, RISCV_SI_FTYPE_SI_SI,           cvbitmanip),
```
Comment (jeremybennett) [inline gcc/config/riscv/corev.def]: This should be `RISCV_USI_FTYPE_USI_USI`

File: gcc/config/riscv/riscv.md [hunk 6]
```diff
@@ -45,6 +45,7 @@
 
   ;; Stack tie
   UNSPEC_TIE
+
```
Comment (jeremybennett) [inline gcc/config/riscv/riscv.md]: Why add a blank line?

File: gcc/doc/extend.texi [hunk 7]
```diff
@@ -21118,6 +21118,59 @@ Generates the @code{cv.macsRN} machine instruction.
 Generates the @code{cv.machhsRN} machine instruction.
 @end deftypefn
 
+These built-in functions are available for the CORE-V bit manipulation machine
+architecture. For more information on CORE-V built-ins, please see
+@uref{https://github.com/openhwgroup/core-v-sw/blob/master/specifications/corev-builtin-spec.md#pulp-bit-manipulation-builtins-for-32-bit-cores}
+
+@deftypefn {Built-in Function} {int32_t} __builtin_riscv_cv_bitmanip_extract (int32_t, uint16_t)
+Generated assembler @code{cv.extract} is the uint16_t operand is a constant. Generated assembler
```
Comment (jeremybennett) [inline gcc/doc/extend.texi]: is -> if (also in all succeeding blocks)

File: gcc/doc/extend.texi [hunk 8]
```diff
@@ -21118,6 +21118,59 @@ Generates the @code{cv.macsRN} machine instruction.
...
+@code{cv.extractr} is the uint16_t operand is a register.
```
Comment (jeremybennett) [inline gcc/doc/extend.texi]: is -> if (also in all succeeding blocks)

**Review by jeremybennett — CHANGES_REQUESTED** [review]
Nice work. A few small changes needed.

File: gcc/config/riscv/corev.def (hunk 3 above)
Comment (MaryBennett) [reply gcc/config/riscv/corev.def]: Updated

File: gcc/config/riscv/corev.def (hunk 5 above)
Comment (MaryBennett) [reply gcc/config/riscv/corev.def]: Updated

File: gcc/doc/extend.texi (hunk 8 above)
Comment (MaryBennett) [reply gcc/doc/extend.texi]: Updated

File: gcc/doc/extend.texi (hunk 7 above)
Comment (MaryBennett) [reply gcc/doc/extend.texi]: Updated

File: gcc/config/riscv/riscv.md (hunk 6 above)
Comment (MaryBennett) [reply gcc/config/riscv/riscv.md]: Updated

File: gcc/config/riscv/corev.def (hunk 4 above)
Comment (MaryBennett) [reply gcc/config/riscv/corev.def]: Updated

File: gcc/config/riscv/corev.def (hunk 2 above)
Comment (MaryBennett) [reply gcc/config/riscv/corev.def]: Updated

File: gcc/config/riscv/constraints.md (hunk 1 above)
Comment (MaryBennett) [reply gcc/config/riscv/constraints.md:85]: GCC stage 1 will produce the following error "Multi-letter constraints with first letter 'C' should have same length".

Comment (jeremybennett) [reply gcc/config/riscv/constraints.md:85]: OK, leave it for now.

**Review by jeremybennett — APPROVED** [review]
This is now good to merge.

**Pull request merged**